src/main/java/org/magicdgs/readtools/utils/read/stats/engine/ProperStatWindowEngine.java

```diff
@@ -0,0 +1,275 @@
+                ProperStatWindowCalculator.createColumnNames(singleStats, pairStats);
+
+        // create the TreeMap with sorting order using the dictionary
+        this.windowsPerContig = new TreeMap<>(Comparator.comparingInt(dict::getSequenceIndex));
+        intervals.stream()
+                .map(i -> new ProperStatWindowCalculator(i, singleStats, pairStats, columnNames))
+                .forEach(swc -> windowsPerContig.compute(swc.getContig(), (k, v) -> {
+                    if (v == null) {
+                        v = new ArrayList<>();
+                    }
+                    v.add(swc);
+                    return v;
+                }));
+
+
+        // create the writer and write first the header
+        this.writer = new PrintStream(stream);
+        this.printAll = printAll;
+        writeHeader();
+    }
+
+    /**
+     * Adds the read to the computation engine.
+     *
+     * <p>Note that the caller is responsible of assesing the following requirements:
+     *
+     * <ul>
+     *     <li>Reads are added in a coordinate sorted ordering.</li>
+     *     <li>Supplementary/Secondary alignments are not passed to this method.</li>
+     * </ul>
+     *
+     * <p>Results are unpredictable otherwise.
+     *
+     * @param read the read to be added.
+     */
+    public void addRead(final GATKRead read) {
+        final List<ProperStatWindowCalculator> wins = flushUpToRead(read);
+        for (final ProperStatWindowCalculator w: wins) {
+            // TODO: performance improvement - should add a check for break the for loop once
+            // TODO: there are no-windows requiring this read
+            // TODO: if using IntervalsSkipList it might be easier by querying needed ones
+            w.addRead(read);
+        }
+    }
+
+    /**
+     * Flushes the queue of windows up to the provided read.
+     *
+     * @param read next read to be added.
+     * @return windows overlapping with the read.
+     */
+    private List<ProperStatWindowCalculator> flushUpToRead(final GATKRead read) {
+        // edge case where the windows are already finalized
+        if (windowsPerContig.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        // store the first key
+        String firstKey = windowsPerContig.firstKey();
+        // only if the read is mapped, otherwise does not have a contig
+        if (!read.isUnmapped()) {
+            logger.debug("Flush up to {} ({} contigs)", read::toString, windowsPerContig::size);
+            final String readContig = read.getContig();
+            // remove the contigs that are before the current read
+            while (!windowsPerContig.isEmpty() && windowsPerContig.comparator().compare(firstKey, readContig) < 0) {
+                logger.debug("Output contig {}", firstKey);
+                windowsPerContig.remove(firstKey).forEach(this::maybeWriteWindow);
+                firstKey = (windowsPerContig.isEmpty()) ? null : windowsPerContig.firstKey();
```

Comment:
to get rid of the null stuff - in the first iteration, the windowsPerContig is never empty and thus:

* if empty, we can break the while loop
* we avoid checking twice if it is empty
* the last check for the null firstKey could be modified for a check of isEmpty

Reply:
Cleaned!